Freezing MenuItem In folder fragment.

Pinankh opened the pull request:
### Fixes UI freeze issue caused by media scanning in FolderFragment.

This PR addresses a freezing issue in the FolderFragment when the user selects the "Scan" item from the options or popup menu. The freezing was caused by the direct invocation of MediaScannerConnection, which performs blocking operations on the main thread.

**Before**

https://github.com/user-attachments/assets/4e1e8ab7-626a-4e43-9959-505f00f7dcad



**After**

https://github.com/user-attachments/assets/de10578a-d475-4588-ad8c-58887f43f468



**🛠️ Changes Made**
To resolve this, I made the following changes:

- Refactored **scanPaths()** and **listPaths()**

Modified these methods to ensure that media scanning operations are performed off the main thread, preventing UI lag or freezing.

- Added **ScanMusicBottomSheet.kt**

A new bottom sheet dialog fragment that provides a more modern and user-friendly interface for scanning folders. This also ensures better UX separation and lifecycle handling.

- Added **ScanViewModel.kt**

Implements a ViewModel to manage scan-related state and lifecycle-aware background processing using Kotlin coroutines or LiveData (depending on existing architecture).

- Added **FolderMediaScannerCompletionListener.java**

A callback interface that handles scan completion logic, improving modularity and ensuring smooth updates after the scan finishes.

**✅ Benefits**

1. Fixes the menu freezing issue when initiating a folder scan.

1. Ensures responsive UI by moving scanning operations off the main thread.

1. Improves code modularity and adherence to MVVM principles.

1. Introduces a modern UI component (BottomSheet) for scanning operations.

**🧪 Testing**

1. Verified that the scan action no longer freezes the menu or UI.

1. Confirmed that media scanning works correctly and updates the library as expected.

1. Checked compatibility with existing FolderFragment behavior and navigation.

**📎 Related Issue**
Issue: "In the FolderFragment, the Scan menu option freezes due to MediaScanner call."

renovate[bot] committed:
fix(deps): update dependency androidx.gridlayout:gridlayout to v1.1.0 (#29)

* fix(deps): update navigation.version to v2.9.0 (#25)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* fix(deps): update lifecycle.version to v2.9.0 (#24)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* fix(deps): update dependency androidx.recyclerview:recyclerview to v1.4.0 (#23)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* fix(deps): update dependency androidx.gridlayout:gridlayout to v1.1.0

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

tomaThomas committed:
Update gradle actions

tomaThomas committed:
Update dependencies

renovate[bot] committed:
chore(deps): update gradle/actions action to v4 (#32)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency gradle to v8.14.2 (#27)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Pinankh committed:
I found the issue in FoldersFragment.kt whenever user tap on Scan the Menu is freezing. So i changed little bit code in scanPaths() Function and Add ScanMusicBottomSheet.kt and ScanViewModel.kt and FolderMediaScannerCompletionListener.java.

tomaThomas commented:
I dont think that radar scanner thing fits to the material design

And this is clearly completely AI generated, you did not even bother to remove the Optional: ... comments, there is logging commented out. Please clean up the code (also make sure you understand it and remove unnecessary stuff) before creating a PR

tomaThomas reviewed (changes requested):
Don't just copy paste from AI

Pinankh commented:
> I dont think that radar scanner thing fits to the material design
> 
> And this is clearly completely AI generated, you did not even bother to remove the Optional: ... comments, there is logging commented out. Please clean up the code (also make sure you understand it and remove unnecessary stuff) before creating a PR

Thaks for your reply Sir, I will Remove the Animation BottomSheet Dailog and The Yes i get some help from AI. I will remove the Unnecessary stuff. And Sorry for this unfixed PR. I am New in Contributing the project. Next time i will be careful.

tomaThomas commented:
A loading indicator would be fine, but I'd stick with a material spinner / sth. More simplistic

Thanks for your contribution

Pinankh committed:
Refactored code as per review: removed comments, logs, and unnecessary animations.

Pinankh commented:
> A loading indicator would be fine, but I'd stick with a material spinner / sth. More simplistic
> 
> Thanks for your contribution

You are always welcome Thomas sir. Okay I added material spinner instead of animated view. Also Refactored code as per review: removed comments, logs, and unnecessary animations. Please review the code and marge. I hope this time i done correctly.

tomaThomas commented on gradle/libs.versions.toml:
```diff
@@ -83,6 +84,7 @@ afollestad-material-dialogs-input = { module = "com.afollestad.material-dialogs:
 afollestad-material-cab = { module = "com.afollestad:material-cab", version.ref = "afollestad_materialCab_version" }
 
 #Retrofit and apos
+lottie = { module = "com.airbnb.android:lottie", version.ref = "lottieVersion" }
```
Why do we need this dependency?

tomaThomas commented on app/src/main/res/values/strings.xml:
```diff
@@ -571,4 +571,18 @@
     <string name="sleep_timer_no_permission">No permission to schedule alarms, please allow Retro Music.</string>
+    <string name="start_scan">Start Scan</string>
+    <string name="close">Close</string>
+    <string name="scanning_folders_files">Scanning Folders And Files...</string>
```
Please use consistent capitalization (i.e. lower case)

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/helper/ScanMusicBottomSheet.kt:
This should be a formatted string (then it can also work for right to left languages)

Pinankh committed:
Refactored code as per review: remove Lottie Dependency from lib From libVersion,make consistent capitalization in strings.xml, And make formatted string so work for right to left languages now.

Pinankh commented:
**### Review the request**
Refactored code as per review: 
1:- Remove Lottie Dependency from lib From libVersion.
2:- Make consistent capitalization in strings.xml, 
3:- Make formatted string so work for right to left languages now.

And Thanks for addressing me this issue on my pull request.

tomaThomas commented on gradle/libs.versions.toml:
```diff
@@ -2,6 +2,7 @@
 #buildscripts dependencies
 advrecyclerview = "1.0.0"
 jetradarmobile_androidSnowfall_version = "1.2.1"
+lottieVersion = "3.4.0"
```
This is not used anymore

tomaThomas commented on app/src/main/res/layout/bottom_sheet_scan_music.xml:
```diff
@@ -0,0 +1,74 @@
+        android:textAppearance="?attr/textAppearanceBodyMedium"
+        android:paddingBottom="16dp"/>
+
+    <com.google.android.material.button.MaterialButton
```
Can we get rid of this button? I think the user cannot do anything else here, so that would reduce the numbe rof clicks necessary

Pinankh replied:
Do you Mean both Start Scan and Close button? Or only Start Scan button?

tomaThomas replied:
I meant the start button, the close button gives the user the option to see the result

Pinankh replied:
Okay You are right there is already user tap on scan menu item. So Start Scan not nedded i got it.

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/misc/FolderMediaScannerCompletionListener.java:
Why are these class variables?

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/misc/FolderMediaScannerCompletionListener.java:
This should come from R.strings. I think we can use the same success message in all cases here; if there is nothing to scan / found it can just say 0 media items found.

Pinankh committed:
Remove Lottie Version, Remove Start Scan button from ScanMusicBottomSheet.kt, Remove Unnecessary variable form FolderMediaScannerCompletionListener.java, And use the same success message in all cases here

Pinankh commented:
Hello Thomas Please check and review i changed and removed as per your suggesttion.

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/misc/FolderMediaScannerCompletionListener.java:
Please use a single formatted string again :)

Pinankh commented:
Hello Thomas i am waiting for your response. Its been one week i didnt hear from you. I hope you are allright,

tomaThomas commented on app/src/main/res/values/strings.xml:
```diff
@@ -571,4 +571,20 @@
     <string name="your_account_data_is_only_used_for_authentication">Your account data is only used for authentication.</string>
     <string name="could_not_write_tags_to_file">Could not write tags to the music file!</string>
     <string name="sleep_timer_no_permission">No permission to schedule alarms, please allow Retro Music.</string>
+    <string name="start_scan">Start scan</string>
```
Do we really need this many strings? I think we can reduce the number, i.e. we don't need messages for errors that will probably never occur (like when can context be null)
Additionally, "context" means nothing to the user, this is a message for developers

Pinankh committed:
use a single formatted string with LTR & RTL.Remove error massage string and some other unnecessary strings

Pinankh commented:
Hello Thomas Thanks for your response. I removed unnecessary Strings and make One string with LTR And RTL support. Please check and review the PR and commite.

tomaThomas commented:
I still did not check the PRon the PC, only on mobile. I hope I find the time this weekend :)

Pinankh commented:
It's okay Thomas take your time.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Thomas Möller ***@***.***>
Sent: Wednesday, July 2, 2025 10:36:09 PM
To: RetroMusicPlayer/RetroMusicPlayer ***@***.***>
Cc: Pinankh Patel ***@***.***>; Author ***@***.***>
Subject: Re: [RetroMusicPlayer/RetroMusicPlayer] Freezing MenuItem In folder fragment. (PR #1811)

[https://avatars.githubusercontent.com/u/17884154?s=20&v=4]tomaThomas left a comment (RetroMusicPlayer/RetroMusicPlayer#1811)<https://github.com/RetroMusicPlayer/RetroMusicPlayer/pull/1811#issuecomment-3028611064>

I still did not check the PRon the PC, only on mobile. I hope I find the time this weekend :)

—
Reply to this email directly, view it on GitHub<https://github.com/RetroMusicPlayer/RetroMusicPlayer/pull/1811#issuecomment-3028611064>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHBZFWTIUJYY5HVVIKIR3N33GQGQDAVCNFSM6AAAAAB7NMDY52VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMRYGYYTCMBWGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/helper/ScanMusicBottomSheet.kt:
Remove this commented out code

Pinankh replied:
okay got it

2376780283 commented:
this func have problem on tablet

![Screenshot_2025-07-14-12-15-17-437_zzh lifeplayer music debug](https://github.com/user-attachments/assets/44ae5efb-3601-4713-8f70-a0994031750b)

tomaThomas commented:
> this func have problem on tablet

Thanks for trying this!

2376780283 commented:
# 建议报告：解决 Scan BottomSheet 在平板横屏状态下无法完全展开的问题导致显示不完全

## 背景说明  
本人是新手开发者，刚开始使用 GitHub 进行项目协作，可能对一些流程和术语还不够熟悉。  
因此，在发现问题并反馈时，交流中可能存在表达不够准确或不完善的地方，敬请各位开发者理解与指正。

## 1. 问题描述  
在平板设备的横屏状态下，RetroMusicPlayer 中扫描音乐文件夹的scan BottomSheet 无法完全展开，导致用户体验受影响。  
- 影响 BottomSheet 的正常使用和界面展示。

## 2. 已尝试的解决方案  
- **方案一**：通过调整 BottomSheet 的高度限制来解决问题。  
- **方案二**：修改 BottomSheet 的布局参数以适应横屏状态。  

然而，经过测试，这些方案未能完全解决在平板横屏状态下 BottomSheet 无法完全展开的问题。

## 3. 解决方案建议  
经过进一步调试和测试，发现必须在横屏状态下**强制设置 BottomSheet 处于完全展开状态**，才能正常使用。  
- 强制展开 BottomSheet 能保证其高度覆盖屏幕，避免显示不全。 
##详细截图 
# 尝试修复后	
![Screenshot_2025-07-14-15-10-49-148_zzh lifeplayer music](https://github.com/user-attachments/assets/73c54865-f215-42c0-a7a5-2f2a81e653a8)

Pinankh commented:
> # 建议报告：解决 Scan BottomSheet 在平板横屏状态下无法完全展开的问题导致显示不完全
> ## 背景说明
> 本人是新手开发者，刚开始使用 GitHub 进行项目协作，可能对一些流程和术语还不够熟悉。 因此，在发现问题并反馈时，交流中可能存在表达不够准确或不完善的地方，敬请各位开发者理解与指正。
> 
> ## 1. 问题描述
> 在平板设备的横屏状态下，RetroMusicPlayer 中扫描音乐文件夹的scan BottomSheet 无法完全展开，导致用户体验受影响。
> 
> * 影响 BottomSheet 的正常使用和界面展示。
> 
> ## 2. 已尝试的解决方案
> * **方案一**：通过调整 BottomSheet 的高度限制来解决问题。
> * **方案二**：修改 BottomSheet 的布局参数以适应横屏状态。
> 
> 然而，经过测试，这些方案未能完全解决在平板横屏状态下 BottomSheet 无法完全展开的问题。
> 
> ## 3. 解决方案建议
> 经过进一步调试和测试，发现必须在横屏状态下**强制设置 BottomSheet 处于完全展开状态**，才能正常使用。
> 
> * 强制展开 BottomSheet 能保证其高度覆盖屏幕，避免显示不全。
>   ##详细截图
> 
> # 尝试修复后
> ![Screenshot_2025-07-14-15-10-49-148_zzh lifeplayer music](https://private-user-images.githubusercontent.com/118289872/465877611-73c54865-f215-42c0-a7a5-2f2a81e653a8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2NjQzMDQsIm5iZiI6MTc1MjY2NDAwNCwicGF0aCI6Ii8xMTgyODk4NzIvNDY1ODc3NjExLTczYzU0ODY1LWYyMTUtNDJjMC1hN2E1LTJmMmE4MWU2NTNhOC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxNlQxMTA2NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYTNkY2E0OGMwMzZkOThjOTUxNmU5MDc5NzVmOTk1NGU5ZjMzZDMzZDI0MzY4N2MwOGU5ODAyNDY2ZjZhOWQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UONRIdQMKa47GUUCAlPkbFiUcIj4NQQG5fwRN0mYELE)

Thanks for your Solution sugession. I appreciate

Pinankh committed:
fixed Scanning BottomSheet problem on Horizontal screen state.

Pinankh commented:
Hiii Thomas Thanks for addressing the issue. i fixed the Scanning BottomSheet problem on Horizontal screen state. Please check the PR. I appreciate your time and effort.

_**This is on mobile Protraid Mode**_
<img width="1084" height="2412" alt="Screenshot_20250716_173743" src="https://github.com/user-attachments/assets/5e038ad6-7f5f-406d-8505-b8a4988d8512" />

**_This is on Tablet Horizontal screen mode_**
<img width="2412" height="1084" alt="Screenshot_20250716_173841" src="https://github.com/user-attachments/assets/a2d9390b-59e6-4e20-81e9-e9e0d1cc4687" />

tomaThomas committed:
Minor cleanup

tomaThomas committed:
Remove unused string

tomaThomas commented:
Sorry for taking me so long, I barely used my PC the last weeks ;)
I did a bit of cleanup (mainly removed unused or redundant translated strings)
If the pipeline succeeds, we can merge the PR :)

tomaThomas committed:
Merge branch 'dev', remote-tracking branch 'upstream' into pinankh

tomaThomas committed:
Fix build
